solver-next/llb/result.go

```diff
@@ -0,0 +1,138 @@
+package llb
+
+import (
+	"bytes"
+	"context"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/moby/buildkit/cache"
+	"github.com/moby/buildkit/cache/contenthash"
+	solver "github.com/moby/buildkit/solver-next"
+	"github.com/moby/buildkit/worker"
+	digest "github.com/opencontainers/go-digest"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+)
+
+func NewWorkerRefResult(ref cache.ImmutableRef, worker worker.Worker) solver.Result {
+	return &workerRefResult{&WorkerRef{ImmutableRef: ref, Worker: worker}}
+}
+
+type WorkerRef struct {
+	ImmutableRef cache.ImmutableRef
+	Worker       worker.Worker
+}
+
+func (wr *WorkerRef) ID() string {
+	return wr.Worker.ID() + "::" + wr.ImmutableRef.ID()
+}
+
+type workerRefResult struct {
+	*WorkerRef
+}
+
+func (r *workerRefResult) Release(ctx context.Context) error {
+	return r.ImmutableRef.Release(ctx)
+}
+
+func (r *workerRefResult) Sys() interface{} {
+	return r.WorkerRef
+}
+
+func NewContentHashFunc(selectors []string) solver.ResultBasedCacheFunc {
+	return func(ctx context.Context, res solver.Result) (digest.Digest, error) {
+		ref, ok := res.Sys().(*WorkerRef)
+		if !ok {
+			return "", errors.Errorf("invalid reference: %T", res)
+		}
+
+		dgsts := make([][]byte, len(selectors))
+
+		eg, ctx := errgroup.WithContext(ctx)
+
+		for i, sel := range selectors {
+			// FIXME(tonistiigi): enabling this parallelization seems to create wrong results for some big inputs(like gobuild)
+			// func(i int) {
+			// 	eg.Go(func() error {
+			dgst, err := contenthash.Checksum(ctx, ref.ImmutableRef, path.Join("/", sel))
+			if err != nil {
+				return "", err
+			}
+			dgsts[i] = []byte(dgst)
+			// return nil
+			// })
+			// }(i)
+		}
+
+		if err := eg.Wait(); err != nil {
+			return "", err
+		}
+
+		return digest.FromBytes(bytes.Join(dgsts, []byte{0})), nil
```

Comment:
What is this `[]byte{0}`?

Reply:
NULL byte separator between selectors. I can use some other character.